File: src/lc_125_valid_palindrome.c
#include <stdbool.h>

char lower_case(char c)
{
    if ((c >= 'A') && (c <= 'Z'))
    {
        //
        int conv_factor = 'a' - 'A';

        c = c + conv_factor;
        return c;
    }
    else
    {
        return c;
    }
}

bool is_alphanumeric(char c)
{
    if ( (c >= 'a') && (c <= 'z') || (c >= 'A') && (c <= 'Z') || (c >= '0') && (c <= '9') )
    {
        return true;    
    }

    return false;

}

int len_string(char * s)
{
    int counter = 0;
    int i       = 0;

    for ( i; s[i] != '\0'; i++)
    {
        counter++;
    }

    return counter;
}

bool isPalindrome(char* s)
{
    int left_index  = 0;
    int right_index = len_string(s) - 1;
    
    while (left_index < right_index)
    {
        while ( (left_index < right_index) && ( ! is_alphanumeric(s[left_index])) )
        {
            left_index++;
        }

        while ( (left_index < right_index) && ( ! is_alphanumeric(s[right_index])) )
        {
            right_index--;
        }

        if ( ! ( lower_case(s[left_index]) == lower_case(s[right_index]) ) )
        {
            return false;
        }

        left_index++;
        right_index--;
    }
    
    return true;
}

File: src/lc_125_valid_palindrome.py
class ValidPalindrome():

    def __init__(self, input_string):
        self.input_string   = input_string

    def is_alphanumeric(self, my_char):
        
        if ( (my_char >= 'a') and (my_char <= 'z') \
            or (my_char >= 'A') and (my_char <= 'Z') \
            or (my_char >= '0') and (my_char <= '9') ) :
            return True
        else:
            return False

        
    def solve(self):

        s = self.input_string

        left_index  = 0
        right_index = len(s) - 1

        #New string with all lower case characters
        new_string = s.lower()

        while( left_index < right_index):

            while ( left_index < right_index) and ( not self.is_alphanumeric(new_string[left_index]) ):
                left_index += 1

            while ( left_index < right_index) and ( not self.is_alphanumeric(new_string[right_index]) ):
                right_index -= 1
            
            if ( not (new_string[left_index] == new_string[right_index] ) ):
                return False
            
            left_index += 1
            right_index -= 1
            
        return True